<!DOCTYPE html>
<!-- 使用Promise.all等待多个Promise -->
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>

</body>
<script src="./utils.js"></script>
<script>
    function getJSON(url) {
        // 创建并返回一个新的Promise对象
        return new Promise((resolve, reject) => {
            // 创建一个XML Httprequest对象
            const request = new XMLHttpRequest();

            // 初始化请求
            request.open("GET", url);

            // 注册一个onload方法，当服务器响应后会被调用
            request.onload = function () {
                try {
                    // 即使服务正常响应也并不意味着一切如期发生
                    // 只有当服务端返回的状态码为200（一切正常）时，
                    // 再使用服务端的返回结果
                    if (this.status === 200) {
                        // 尝试解析JSON字符串，倘若解析成功则执行resolve，并将解析后的对象作为参数传入
                        resolve(JSON.parse(this.response));
                    } else {
                        reject(this.status + " " + this.statusText);
                    }
                } catch (e) {
                    // 如果服务器返回了不同的状态码或是解析JSON字符串时发生异常，则reject
                    reject(e.message);
                }
            }

            request.onerror = function () {
                // 如果和服务器端通信过程中发生了错误，则reject
                reject(this.status + " " + this.statusText);
            }

            request.send();
        })
    }


    Promise.all([
        getJSON("http://127.0.0.1:5500/data/ninjas.json"),
        getJSON("http://127.0.0.1:5500/data/mapInfo.json"),
        getJSON("http://127.0.0.1:5500/data/plan.json")
    ]).then(results => {
        // 结果将是所有promise成功值组成的数组，其每一项对应promise数组的每一项
        const ninjas = results[0], mapInfo = results[1], plan = results[2];


        // Promise.all接受一个promise数组，并创建一个新的promise对象
        // 当所有的promise成功时该promise为成功状态，否则失败
        assert(ninjas !== undefined
            && mapInfo !== undefined
            && plan !== undefined,
            "The plan is ready to be set in motion!");
    }).catch(error => {
        fail("A problem in carrying out our plan!");
    })

    // 如果只关注第一个成功或者失败的promise，使用Promise.race()
</script>

</html>